astropy/table/column.py

```diff
@@ -790,8 +793,20 @@ def __bytes__(self):
         __str__ = __bytes__
 
     def _check_string_truncate(self, value):
+        """
+        Emit a warning if any elements of ``value`` will be truncated when
+        ``value`` is assigned to self.
+        """
+        # Convert input ``value`` to the string dtype of this column
         value = np.asanyarray(value, dtype=self.dtype.type)
-        if value.dtype.itemsize > self.dtype.itemsize:
+
+        # Compute itemsize for ``value`` as the max *actual length* times the
+        # itemsize for a length=1 version of the string type.  Remember itemsize
+        # is the total length in bytes for a single item.
+        value_str_len = np.char.str_len(value).max()
+        value_itemsize = value_str_len * np.array('', dtype=value.dtype.type).itemsize
```

Comment:
Can we always be sure that `'1'` (or any ASCII character) will be part of any future arbitrary string dtype?  That was my motivation for using an empty string.  I thought about your second solution as well.  Even though I already used that for making the warning message, it still strikes me as bad form to parse the `.str` attribute like that for actual code logic.  Agreed that there really should be a better way!

Anyway, I'll do something to avoid using the zero-length string.